File: src/main/java/org/manapart/microMods/HorseStats.java
package org.manapart.microMods;

import net.minecraft.entity.SharedMonsterAttributes;
import net.minecraft.entity.passive.horse.AbstractHorseEntity;
import net.minecraft.entity.passive.horse.LlamaEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.Items;
import net.minecraft.util.text.TextFormatting;
import net.minecraft.util.text.TranslationTextComponent;
import net.minecraftforge.event.entity.player.PlayerInteractEvent;

public class HorseStats {

    public void getStats(PlayerInteractEvent.EntityInteract event) {
        if (event.getPlayer() != null && event.getWorld() != null && event.getTarget() instanceof AbstractHorseEntity) {
            PlayerEntity player = event.getPlayer();
            AbstractHorseEntity horseEntity = (AbstractHorseEntity) event.getTarget();
            if (player.isCrouching() && player.getHeldItemMainhand().getItem() == Items.COMPASS) {
                event.setCanceled(true);
                displayHorseStats(player, horseEntity);
            }
        }
    }

    private void displayHorseStats(PlayerEntity player, AbstractHorseEntity horseEntity) {
        double health = horseEntity.getAttribute(SharedMonsterAttributes.MAX_HEALTH).getValue();
        double speed = horseEntity.getAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).getValue();
        double jump = horseEntity.getHorseJumpStrength();

        jump = -0.1817584952 * Math.pow(jump, 3) + 3.689713992 * Math.pow(jump, 2) + 2.128599134 * jump - 0.343930367; // https://minecraft.gamepedia.com/Horse
        speed = speed * 43;

        TextFormatting colourHealth = getHealthColor(health);
        TextFormatting colourSpeed = getSpeedColor(speed);
        TextFormatting colourJump = getJumpColor(jump);


        if (horseEntity instanceof LlamaEntity) {
            displayLlamaMessage(player, (LlamaEntity) horseEntity, colourHealth, colourSpeed, health, speed);
        } else {
            displayHorseMessage(player, colourSpeed, colourJump, health, speed, jump, colourHealth);
        }
    }

    private TextFormatting getHealthColor(double health) {
        TextFormatting color = TextFormatting.WHITE;
        if (health <= 20) {
            color = TextFormatting.GRAY;
        } else if (health <= 23) {
            color = TextFormatting.WHITE;
        } else if (health <= 26) {
            color = TextFormatting.YELLOW;
        } else if (health <= 29) {
            color = TextFormatting.AQUA;
        } else if (health <= Double.MAX_VALUE) {
            color = TextFormatting.LIGHT_PURPLE;
        }
        return color;
    }

    private TextFormatting getSpeedColor(double speed) {
        TextFormatting color;
        if (speed <= 7) {
            color = TextFormatting.GRAY;
        } else if (speed <= 9) {
            color = TextFormatting.WHITE;
        } else if (speed <= 11) {
            color = TextFormatting.YELLOW;
        } else if (speed <= 13) {
            color = TextFormatting.AQUA;
        } else if (speed <= Double.MAX_VALUE) {
            color = TextFormatting.LIGHT_PURPLE;
        } else {
            color = TextFormatting.LIGHT_PURPLE;
        }
        return color;
    }

    private TextFormatting getJumpColor(double jump) {
        TextFormatting color = TextFormatting.WHITE;
        if (jump <= 1.50) {
            color = TextFormatting.GRAY;
        } else if (jump <= 2.0) {
            color = TextFormatting.WHITE;
        } else if (jump <= 3.0) {
            color = TextFormatting.YELLOW;
        } else if (jump <= 4.0) {
            color = TextFormatting.AQUA;
        } else if (jump <= Double.MAX_VALUE) {
            color = TextFormatting.LIGHT_PURPLE;
        }
        return color;
    }

    private TextFormatting getSlotsColor(double slots) {
        TextFormatting color = TextFormatting.WHITE;
        if (slots <= 3) {
            color = TextFormatting.GRAY;
        } else if (slots <= 6) {
            color = TextFormatting.WHITE;
        } else if (slots <= 9) {
            color = TextFormatting.YELLOW;
        } else if (slots <= 12) {
            color = TextFormatting.AQUA;
        } else if (slots <= Double.MAX_VALUE) {
            color = TextFormatting.LIGHT_PURPLE;
        }
        return color;
    }

    private void displayLlamaMessage(PlayerEntity player, LlamaEntity horseEntity, TextFormatting colourHealth, TextFormatting colourSpeed, double health, double speed) {
        double slots = horseEntity.getStrength();

        slots = slots * 3;

        TextFormatting colourSlots = getSlotsColor(slots);

        player.sendStatusMessage(new TranslationTextComponent(String.format("%sHealth: %.0f %sSpeed: %.1f %sChest Slots: %.0f", colourHealth, health, colourSpeed, speed, colourSlots, slots)), true);
    }

    private void displayHorseMessage(PlayerEntity player, TextFormatting colourSpeed, TextFormatting colourJump, double health, double speed, double jump, TextFormatting colourHealth) {
        player.sendStatusMessage(new TranslationTextComponent(String.format("%sHealth: %.0f %sSpeed: %.1f %sJump Height: %.1f", colourHealth, health, colourSpeed, speed, colourJump, jump)), true);
    }

}

File: src/main/java/org/manapart/microMods/HorseStats.kt
package org.manapart.microMods

import net.minecraft.ChatFormatting
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.MutableComponent
import net.minecraft.world.entity.ai.attributes.Attributes
import net.minecraft.world.entity.animal.horse.AbstractHorse
import net.minecraft.world.entity.animal.horse.Llama
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.Items
import net.minecraftforge.event.entity.player.PlayerInteractEvent
import java.awt.TextComponent
import kotlin.math.pow


class HorseStats {
    fun getStats(event: PlayerInteractEvent.EntityInteract) {
        if (event.entity != null && event.level != null && event.target is AbstractHorse) {
            val player = event.entity
            val horseEntity = event.target as AbstractHorse
            if (player.isCrouching && player.mainHandItem.item === Items.COMPASS) {
                event.isCanceled = true
                displayHorseStats(player, horseEntity)
            }
        }
    }

    private fun displayHorseStats(player: Player, horseEntity: AbstractHorse) {
        val health = horseEntity.getAttributeValue(Attributes.MAX_HEALTH)
        var speed = horseEntity.getAttributeValue(Attributes.MOVEMENT_SPEED)
        var jump = horseEntity.customJump
        jump = -0.1817584952 * jump.pow(3.0) + 3.689713992 * jump.pow(2.0) + 2.128599134 * jump - 0.343930367 // https://minecraft.gamepedia.com/Horse
        speed *= 43
        val colourHealth = getHealthColor(health)
        val colourSpeed = getSpeedColor(speed)
        val colourJump = getJumpColor(jump)
        if (horseEntity is Llama) {
            displayLlamaMessage(player, horseEntity, colourHealth, colourSpeed, health, speed)
        } else {
            displayHorseMessage(player, colourSpeed, colourJump, health, speed, jump, colourHealth)
        }
    }

    private fun getHealthColor(health: Double): ChatFormatting = getColor(health, listOf(20, 23, 26, 29))
    private fun getSpeedColor(speed: Double): ChatFormatting = getColor(speed, listOf(7, 9, 11, 13))
    private fun getJumpColor(jump: Double): ChatFormatting = getColor(jump, listOf(1, 2, 3, 4))
    private fun getSlotsColor(slots: Double): ChatFormatting = getColor(slots, listOf(3, 6, 9, 12))


    private fun getColor(amount: Double, options: List<Int>): ChatFormatting {
        val intAmount = amount.toInt()
        return when {
            intAmount <= options[0] -> ChatFormatting.GRAY
            intAmount <= options[1] -> ChatFormatting.WHITE
            intAmount <= options[2] -> ChatFormatting.YELLOW
            intAmount <= options[3] -> ChatFormatting.AQUA
            else -> ChatFormatting.LIGHT_PURPLE
        }
    }

    private fun displayLlamaMessage(player: Player, horseEntity: Llama, colourHealth: ChatFormatting, colourSpeed: ChatFormatting, health: Double, speed: Double) {
        val slots = horseEntity.strength.toDouble() * 3
        val colourSlots = getSlotsColor(slots)
        val message = Component.literal("")
            .append("Health: %.0f".formatted(health, colourHealth))
            .append(" Speed: %.1f".formatted(speed, colourSpeed))
            .append(" Chest Slots: %.1f".formatted(slots, colourSlots))

        player.displayClientMessage(message, true)
    }

    private fun displayHorseMessage(player: Player, colourSpeed: ChatFormatting, colourJump: ChatFormatting, health: Double, speed: Double, jump: Double, colourHealth: ChatFormatting) {
        val message = Component.literal("")
            .append("Health: %.0f".formatted(health, colourHealth))
            .append(" Speed: %.1f".formatted(speed, colourSpeed))
            .append(" Jump Height: %.1f".formatted(jump, colourJump))

        player.displayClientMessage(message, true)
    }

    private fun String.formatted(amount: Double, color: ChatFormatting): MutableComponent {
        return Component.literal(String.format(this, amount)).withStyle(color)
    }
}